include/drivers/pinctrl/stm32.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * Public APIs for pin control drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_STM32_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_STM32_H_
+
+/**
+ * @brief Pin Controller Interface (STM32)
+ * @defgroup pinctrl_interface_nrf Pin Controller Interface
+ * @ingroup pinctrl_interface
+ * @{
+ */
+
+#include <devicetree.h>
+#include <dt-bindings/pinctrl/stm32-pinctrl.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Type for STM32 pin. */
+typedef struct pinctrl_pin {
+	uint32_t pinmux;
+	uint32_t pincfg;
+} pinctrl_pin_t;
+
+/**
+ * @brief Utility macro to initialize pincfg field.
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_STM32_PINCFG_INIT(node_id, state_prop, idx)		       \
+	(((STM32_NO_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_disable))     \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PULL_UP *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_up))     \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PULL_DOWN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, bias_pull_down))   \
+	   << STM32_PUPDR_SHIFT) |					       \
+	 ((STM32_PUSH_PULL *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_push_pull))  \
+	   << STM32_OTYPER_SHIFT) |					       \
+	 ((STM32_OPEN_DRAIN *						       \
+	   DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, drive_open_drain)) \
+	   << STM32_OTYPER_SHIFT) |					       \
+	 (DT_ENUM_IDX(DT_PROP_BY_IDX(node_id, state_prop, idx), slew_rate)     \
+	   << STM32_OSPEEDR_SHIFT))
+
+/**
+ * @brief Utility macro to initialize each pin.
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_PIN_INIT(node_id, state_prop, idx)			       \
+	{ .pinmux = DT_PROP_BY_PHANDLE_IDX(node_id, state_prop, idx, pinmux),  \
+	  .pincfg = Z_PINCTRL_STM32_PINCFG_INIT(node_id, state_prop, idx) },
+
+/**
+ * @brief Utility macro to initialize pins.
+ *
+ * @param node_id Node identifier.
+ * @param state_idx State index.
+ */
+#define Z_PINCTRL_PINS_INIT(state_idx, node_id)				       \
+	{DT_FOREACH_PROP_ELEM(node_id, pinctrl_ ## state_idx,		       \
```

Comment:
@mbolivar-nordic `##` here, we probably need to extend pinctrl DT API to cover that

Reply:
> @mbolivar-nordic `##` here, we probably need to extend pinctrl DT API to cover that

@gmarull DT_FOREACH_PINCTRL_ELEM(node_id, state_idx, ...)?

Reply:
@mbolivar-nordic yes, sounds good